superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h

```diff
@@ -0,0 +1,254 @@
+    std::vector<int> &get_padA() { return padA_; }
+    std::vector<int> &get_filterStrideA() { return filterStrideA_; }
+    std::vector<int> &get_dilationA() { return dilationA_; }
+    cudnnConvolutionMode_t &get_mode() { return mode_; }
+    bool get_use_tensor_core() { return use_tensor_core_; }
+    cudnnDataType_t &get_input_type() { return input_type_; }
+    cudnnDataType_t &get_conv_type() { return input_type_; }
+    std::string &get_name() { return name; }
+    cudnn_function_name_enum get_e_name() { return e_name; }
+    std::string &get_str() { return function_str_; }
+    /**
+     * @brief Convert name string to enum name
+     * @return cudnn_function_name_enum
+     */
+    cudnn_function_name_enum name2enum() {
+        auto it = cudnn_function_name_string.find(this->name);
+        if (it != cudnn_function_name_string.end()) {
+            this->e_name = it->second;
+            return e_name;
+        } else {
+            throw "ERROR: invalid input function name";
+        }
+    }
+};
+
+/**
+ * @brief Generation of cudnn functions' params and run the benchmark of this function
+ *
+ * @tparam T1 the type of TensorDescriptor
+ * @tparam T2 the type of ConvolutionDescriptor
+ */
+template <typename T1, typename T2> class CudnnFunction : public CudnnConfig {
+  protected:
+    cudnnHandle_t cudnn_handle;
+    TensorDescriptorNd<T1> x_desc_;
+    FilterDescriptorNd<T1> w_desc_;
+    ConvolutionDescriptor<T2> conv_desc_;
+    TensorDescriptorNd<T1> h_desc_;
+    size_t fwd_workspace_size_;
+    float *fwd_workspace_;
+    T1 *x, *filter, *h;
+    const float alpha_ = 1.f;
+    const float beta_ = 0.f;
+
+    /**
+     * @brief Malloc cuda memory and fill in value for data params used in the cudnn function
+     */
+    void prepare_input();
+    /**
+     * @brief Generate some params used in the cudnn function
+     */
+    void prepare_for_function();
+    /**
+     * @brief Get and set convolution algorithm and workspace size used in cudnn convolution functions
+     */
+    virtual void get_workspace_size() {}
+    /**
+     * @brief launch the kernel/function
+     */
+    virtual void kernel_entry() {}
+
+  public:
+    /**
+     * @brief Construct a new Cudnn Function object according to a CudnnConfig object, including initialization for
+     * cudnn handle and curand
+     *
+     * @param config a CudnnConfig object including configuration of the params to the cudnn function
+     */
+    CudnnFunction(CudnnConfig &config) : CudnnConfig(config) {
+        // Init cudnn handle and device
+        cuda_init(&this->cudnn_handle);
+    }
+    /**
+     * @brief Destroy the Cudnn Function object, including free cuda memory and handle of cudnn and curand
+     */
+    virtual ~CudnnFunction() {
+        // free context and memory
+        CUDA_SAFE_CALL(cudaFree(x));
+        CUDA_SAFE_CALL(cudaFree(filter));
+        CUDA_SAFE_CALL(cudaFree(h));
+        cuda_free(&this->cudnn_handle);
```

Comment:
what's the difference between `cudaFree` and `cuda_free`?

Reply:
cudaFree is a api to free the cuda memory of a data pointer, cuda_free is a self-defined function to destroy cudnn handle

Reply:
rename to cudnn_handle_free